### PR TITLE
Drop PyTorch 1.7 support

### LINKED ISSUE #12161: Drop PyTorch 1.7 support
## Proposed refactor

With the release of PL 1.6, we will support PyTorch 1.8 to 1.11 (4 version window)

### Pitch

Do this in 2 steps:

1. Drop testing from CI (important before release)
2. Remove code for its support. (optional before release)
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @carmocca @akihironitta @borda

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #12161

### Does your PR introduce any breaking changes? If yes, please list them.

Drops PyTorch 1.7 support from our codebase

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @justusschock @awaelchli @rohitgr7 @borda

### CODE REVIEW

File: pytorch_lightning/accelerators/gpu.py [hunk 1]
```diff
@@ -99,7 +96,7 @@ def name() -> str:
         return "gpu"
 
 
-def get_nvidia_gpu_stats(device: _DEVICE) -> Dict[str, float]:
+def get_nvidia_gpu_stats(device: _DEVICE) -> Dict[str, float]:  # pragma: no-cover
```
Comment (carmocca) [inline pytorch_lightning/accelerators/gpu.py:99]: Unused, but would need a deprecation

File: pytorch_lightning/utilities/imports.py [hunk 2]
```diff
@@ -89,7 +89,6 @@ def _compare_version(package: str, op: Callable, version: str, use_base_version:
 
 _IS_WINDOWS = platform.system() == "Windows"
 _IS_INTERACTIVE = hasattr(sys, "ps1")  # https://stackoverflow.com/a/64523765
-_TORCH_GREATER_EQUAL_1_8 = _compare_version("torch", operator.ge, "1.8.0")
 _TORCH_GREATER_EQUAL_1_8_1 = _compare_version("torch", operator.ge, "1.8.1")
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:92]: Left this one in case some soul is installing exactly 1.8.0

File: tests/checkpointing/test_checkpoint_callback_frequency.py [hunk 3]
```diff
@@ -87,7 +87,7 @@ def training_step(self, batch, batch_idx):
 
 
 @mock.patch("torch.save")
-@RunIf(standalone=True, min_gpus=2)
+@RunIf(min_gpus=2, standalone=True)
```
Comment (carmocca) [inline tests/checkpointing/test_checkpoint_callback_frequency.py:90]: For some reason, PyCharm decided to touch the order while using its refactoring tool...

If this bothers reviewers, I can undo those changes.

File: pytorch_lightning/strategies/ddp.py [hunk 4]
```diff
@@ -209,9 +211,7 @@ def _register_ddp_hooks(self) -> None:
         log.detail(f"{self.__class__.__name__}: registering ddp hooks")
         # In 1.8, DDP communication hooks only work with NCCL backend and SPSD (single process single device) mode
         # Since 1.9, DDP communication hooks can work on all backends.
-        if _TORCH_GREATER_EQUAL_1_9 or (
-            _TORCH_GREATER_EQUAL_1_8 and self.root_device.type == "cuda" and self._is_single_process_single_device
-        ):
+        if _TORCH_GREATER_EQUAL_1_9 or self.root_device.type == "cuda" and self._is_single_process_single_device:
```
Comment (krshrimali) [inline pytorch_lightning/strategies/ddp.py:214]: nit:

```suggestion
        if _TORCH_GREATER_EQUAL_1_9 or (self.root_device.type == "cuda" and self._is_single_process_single_device):
```

File: tests/models/test_horovod.py [hunk 5]
```diff
@@ -371,7 +371,7 @@ def training_epoch_end(self, outputs) -> None:
 
 # todo: need to be fixed :]
 @pytest.mark.skip(reason="TODO: CI agent.jobstatus=Succeeded: Permission denied")
-@RunIf(skip_windows=True, horovod=True, num_gpus=2)
+@RunIf(horovod=True, skip_windows=True, num_gpus=2)
```
Comment (Borda) [inline tests/models/test_horovod.py:374]: just curious, is there any special reason for this ordering change?
without it the diff would be much smaller and easier to digest... :rabbit:

**Review by Borda — APPROVED** [review]

File: tests/checkpointing/test_checkpoint_callback_frequency.py (hunk 3 above)
Comment (Borda) [reply tests/checkpointing/test_checkpoint_callback_frequency.py:90]: it is just a distraction from the real change... :(

File: tests/models/test_horovod.py (hunk 5 above)
Comment (Borda) [reply tests/models/test_horovod.py:374]: found answer in https://github.com/PyTorchLightning/pytorch-lightning/pull/12432#discussion_r833847774

File: pytorch_lightning/accelerators/gpu.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/gpu.py:99]: this utility is a bit more convenient in terms of the logs that it generates. The one from PT creates a lot of logs that are hard to interpret. I saw a user asking to allow them to somehow differentiate between these 2, either on issues/slack somewhere.

**Review by rohitgr7 — APPROVED** [review]

**Review by krshrimali — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 6]
```diff
@@ -220,7 +220,6 @@ def _worker_check(self, dataloader: DataLoader, name: str) -> None:
 
         # ddp_spawn + num_workers > 0 don't mix! tell the user
         if dataloader.num_workers > 0 and using_spawn:
-            # checks for the attr persistent_workers available in pytorch >= 1.7
             if hasattr(dataloader, "persistent_workers"):
```
Comment (akihironitta) [inline pytorch_lightning/trainer/connectors/data_connector.py:223]: Do we still need to check if dataloader has persistent_workers? (I think the check should've been removed when dropping PyTorch 1.6 support.)
```suggestion
```

https://pytorch.org/docs/1.6.0/data.html#torch.utils.data.DataLoader
https://pytorch.org/docs/1.7.0/data.html#torch.utils.data.DataLoader
https://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 7]
```diff
@@ -238,7 +237,6 @@ def _worker_check(self, dataloader: DataLoader, name: str) -> None:
                 )
 
         elif dataloader.num_workers == 0 and using_spawn:
-            # checks for the attr persistent_workers available in pytorch >= 1.7
             if hasattr(dataloader, "persistent_workers"):
```
Comment (akihironitta) [inline pytorch_lightning/trainer/connectors/data_connector.py:240]: Same as above.
```suggestion
```

**Review by akihironitta — APPROVED** [review]
Can we remove 1.7.* from our script as well?

https://github.com/PyTorchLightning/pytorch-lightning/blob/28dac0c8d9bba4331793c932694eda4eab0bb94e/requirements/adjust-versions.py#L18-L19

**Review by awaelchli — APPROVED** [review]
